pkg/controller/nodes/task/backoff/handler.go

```diff
@@ -0,0 +1,219 @@
+
+func (r *ComputeResourceCeilings) inf() resource.Quantity {
+	// A hack to represent RESOURCE_MAX
+	return resource.MustParse("1Ei")
+}
+
+type ComputeResourceAwareBackOffHandler struct {
+	*SimpleBackOffBlocker
+	*ComputeResourceCeilings
+}
+
+func (h *ComputeResourceAwareBackOffHandler) IsActive() bool {
+	return h.BackOffBaseSecond != 0
+}
+
+// Act based on current backoff interval and set the next one accordingly
+func (h *ComputeResourceAwareBackOffHandler) Handle(ctx context.Context, operation func() error, requestedResourceList v1.ResourceList) error {
+
+	// Pseudo code:
+	// If the backoff is inactive => we should just go ahead and execute the operation(), and handle the error properly
+	//		If operation() fails because of resource => lower the ceiling
+	//		Else we return whatever the result is
+	//
+	// Else if the backoff is active => we should reduce the number of calls to the API server in this case
+	//		If resource is lower than the ceiling => We should try the operation().
+	//			If operation() fails because of the lack of resource, we will lower the ceiling
+	//          Else we return whatever the operation() returns
+	//      Else => we block the operation(), which is where the main improvement comes from
+
+	now := h.Clock.Now()
+	isBlocking := h.SimpleBackOffBlocker.isBlocking(now)
+	isTryable := h.ComputeResourceCeilings.isEligible(requestedResourceList)
+	if !h.IsActive() {
+		return operation()
+	} else if !isBlocking || isTryable {
+		err := operation()
+		if err == nil {
+			logger.Infof(ctx, "The operation was attempted and finished without an error\n")
+			h.SimpleBackOffBlocker.reset()
+			h.ComputeResourceCeilings.resetAll()
+			return nil
+		}
+
+		if IsResourceQuotaExceeded(err) {
+			if !isBlocking {
+				// if the backOffBlocker is not blocking and we are still encountering insufficient resource issue,
+				// we should increase the exponent in the backoff and update the NextEligibleTime
+
+				backOffDuration := h.SimpleBackOffBlocker.backOff()
+				logger.Infof(ctx, "The operation was attempted because the back-off handler is not blocking, but failed due to "+
+					"insufficient resource (backing off for a duration of [%v] to timestamp [%v])\n", backOffDuration, h.SimpleBackOffBlocker.NextEligibleTime)
+			} else if isTryable {
+				// When lowering the ceiling, we only want to lower the ceiling that actually needs to be lowered.
+				// For example, if the creation of a pod requiring X cpus and Y memory got rejected because of
+				// 	insufficient memory, we should only lower the ceiling of memory to Y, without touching the cpu ceiling
+
+				logger.Infof(ctx, "The operation was attempted because the resource requested is lower than the ceilings, "+
+					"but failed due to insufficient resource (the next eligible time remains unchanged [%v]). The requests are "+
+					"[%v]. The ceilings are [%v]\n", h.SimpleBackOffBlocker.NextEligibleTime, requestedResourceList, h.computeResourceCeilings)
+			} else {
```

Comment:
Is this ever possible? maybe just `if !isBlocking` and `else` are enough given the `if !isBlocking || isTryable` ?